server/src/main/java/org/elasticsearch/index/snapshots/blobstore/SlicedInputStream.java

```diff
@@ -98,6 +98,24 @@ public final int read(byte[] buffer, int offset, int length) throws IOException
         return read;
     }
 
+    @Override
+    public long skip(long n) throws IOException {
+        long remaining = n;
+        while (remaining > 0) {
+            final InputStream stream = currentStream();
+            if (stream == null) {
+                break;
+            }
+            final long skipped = stream.skip(remaining);
+            currentSliceOffset += skipped;
+            if (skipped == 0) {
```

Comment:
Hmm `skipped == 0` doesn't necessarily mean we reached end-of-stream. If we haven't quite skipped enough bytes we should try and `read()` one more byte and only move to the next stream if that returns `-1`.

I think we also shouldn't keep retrying the `skip()` if it falls short (unless it reached the end of the stream). It's up to the caller to check the return value and retry if needed.

Reply:
Ah good point! Added a read() to ensure we're reaching stream EOF.

> I think we also shouldn't keep retrying the skip() if it falls short (unless it reached the end of the stream). It's up to the caller to check the return value and retry if needed.

I'm split on this. Certainly the caller should retry, but I also see implementations with loops. E.g., the [default JDK implementation](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/io/InputStream.java#L540) has a loop over `read()`, instead of a single call to `read()`. I also feel it's "safer" in case the caller forgets the loop.

Reply:
Yeah in practice most impls of `skip` do try and skip all the way, just seems like if one doesn't then it might be useful to propagate that to the caller.

I don't think we want to check `skipped == 0` at all tbh - that involves an extra iteration at the end of each stream. Can we just check for EOF if `skipped < remaining`?

Reply:
Yup, totally makes it better, done!